doc/linecard_hotswap/LineCard_hotSwap_HLD.md

```diff
@@ -0,0 +1,406 @@
+[Figure 8: Reapply Configuration flow](./f8.jpg)  
+
+## 1 Introduction and Scope<a name="introduction"></a>
+
+A modern network switch in data center must be a high fault tolerance equipment which consists of many swappable hardware components (e.g., PSU/FAN/QSFP). 
+For stability, it is not allowed to reboot the device after hot-swapping hardware components. The swappable design is very important for data center to minimize the impact of failures on certain hardware components on the switch. 
+This HLD provides a generic software mechanism of line card hot swapping, which includes the architecture, functional flow and CLI command.
+
+## 2 Requirements<a name="requirements"></a>
+
+### 2.1 Functional Requirements<a name="functionality"></a>
+
+**This feature will support the following functionality:**
+1.	A chassis consists of swappable line cards module.
+2.	Each line card consists of multiple external PHY ASIC. The external PHYs on line card must be the same.
+3.	Each line card can be
+    *   plug in
+    *	unplug out
+4.	Plug in
+    *	Event log with card index.
+    *	Automatically apply the configuration for external PHY on line card
+5.	Unplug out
+    *	Event log with card index.
+    *	Both system side and line side are link down.
+    *	User still can configure the port, but the configuration will not take effect until the line card is plugged in again.
+    *	This unplug-out event should not impact all other external PHYs' traffic.
+
+
+### 2.2 Configuration and Management Requirements<a name="configuration-and-management"></a>
+
+Provide CLI commands to display line card status.
+
+## 3 Design<a name="design"></a>
+
+### 3.1 Architecture<a name="architecture"></a>
+The architecture and module dependency are shown in the following figure.
+
+![Line Card Hot Swap High Level Architecture](./f1.jpg "Figure 1: Line Card Hot Swap High Level Architecture")
+
+###### Figure 1: Line Card Hot Swap High Level Architecture
+
+Figure 1 shows the integration of line card hot swap design and SONiC architecture. SAI layer needs to receive the line card state change event from line card driver. In the design, SAI always listens to line card change event. The line card kernel module is in charge of detecting line card status. After boot up, this kernel module should be inserted automatically.
+
+
+### 3.2 OrchAgent Changes<a name="orchagent-changes"></a>
+
+OrchAgent will register an event callback function to publish the line card hot swapping events to DB and PortsOrch will be responsible for processing the line card events.
+
+### 3.3 Syncd Changes<a name="syncd-changes"></a>
+
+A new table, Port State Table, is created in syncd to record port present status. When user configure a port via SAI, syncd will check the port state in Port State Table. If the port is not present, the configuration will not be applied on that port and return SUCCESS.
+
+
+## 3.4 DB Changes<a name="db-changes"></a>
+
```

Comment:
How can I get the line-card no. from a port? Is there any table in config db or state db that indicates that? Or it should be extracted from port name?
Thanks.

Reply:
Actually, We have a new design which based on Gearbox. In the new design, we may get the line card no. from the PHY profile.

Reply:
where can I find it?

Reply:
@stephenxs  Please refer the new commit b9e088a

Reply:
Thanks. Will check it.